xbmc/utils/StringUtils.cpp

```diff
@@ -1134,8 +1134,10 @@ int64_t StringUtils::AlphaNumericCompare(const wchar_t* left, const wchar_t* rig
       return 1;
     if (lsym && rsym)
     {
-      if (lc != rc)
-        return lc - rc;
```

Comment:
Can you post the values of `lc` and `rc` for when this fails?

Reply:
Did some more digging and here is current conclusion.

lc: ' ' 32 0x20
rc: '_' 95 0x5f

the subtracted value of lc-rc is -63 0xffffffc1

size of wchar_t is 4 bytes
size of int64_t is 8 bytes

passing wchar_t as int64_t doesn't work because upper 4 bytes remains zero and value is still positive

doesn't working
	return lc - rc;
	return (int64_t) (lc - rc);
working
	return (int) (lc - rc);
	return (int64_t) lc - (int64_t) rc;

Maybe it is compiler issue?

I also have one small c sample program demonstrating the issue.

Reply:
The problem is sign difference of wchar_t:
```
on x86_64:    #define __WCHAR_TYPE__ int
on ARM:       #define __WCHAR_TYPE__ unsigned int
```

Reply:
Okay thanks for the investigation.

It looks to me like `static_cast<int64_t>(lc - rc)` would be the ticket.

Reply:
Actually no. You can see above that casting the result doesn't work. Because the result is unsigned and doesn't know the negative values.

Reply:
`static_cast<int64_t>(lc) - static_cast<int64_t>(rc)` would work (so the subtraction is done on signed quantities).

Reply:
Yes, this one works. Should I change only return value then and use casting?

And should I also change this https://github.com/xbmc/xbmc/blob/7523e8d04c1c6c219535035e928840bcb48df386/xbmc/utils/StringUtils.cpp#L1348?
I see as same problem but don't know how to test it.

Reply:
Yes I think you should change the other location too. It also looks wrong.

If the function is intended to return -ve, 0, +ve to indicate ordering of two items, then any differences returned must be on signed values.

(the similar `return (nKey1 - nKey2)` paths are fine, as they are already signed).

I don't mind whether the fix is casting to signed, then subtracting or the more long winded compare and return -1, 0, 1. Both should work fine.

Reply:
Yes, both works. But I would rather use some preferred way Kodi is using.
Will change to casting and later I can change to whatever would be better.

Reply:
I would also be in favour of changing lc and rc to a signed integer rather than wchar_t